// @wordpress/block-editor/build-module/components/inner-blocks/index.native.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import { createElement, Fragment } from "@wordpress/element";

function _createSuper(Derived) { return function () { var Super = _getPrototypeOf(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

/**
 * External dependencies
 */
import { pick, isEqual } from 'lodash';
/**
 * WordPress dependencies
 */

import { Component } from '@wordpress/element';
import { withSelect, withDispatch } from '@wordpress/data';
import { synchronizeBlocksWithTemplate as _synchronizeBlocksWithTemplate, withBlockContentContext } from '@wordpress/blocks';
import isShallowEqual from '@wordpress/is-shallow-equal';
import { compose } from '@wordpress/compose';
/**
 * Internal dependencies
 */

import ButtonBlockAppender from './button-block-appender';
import DefaultBlockAppender from './default-block-appender';
import BlockList from '../block-list';
import { withBlockEditContext } from '../block-edit/context';

var InnerBlocks = /*#__PURE__*/function (_Component) {
  _inherits(InnerBlocks, _Component);

  var _super = _createSuper(InnerBlocks);

  function InnerBlocks() {
    var _this;

    _classCallCheck(this, InnerBlocks);

    _this = _super.apply(this, arguments);
    _this.state = {
      templateInProcess: !!_this.props.template
    };

    _this.updateNestedSettings();

    return _this;
  }

  _createClass(InnerBlocks, [{
    key: "getTemplateLock",
    value: function getTemplateLock() {
      var _this$props = this.props,
          templateLock = _this$props.templateLock,
          parentLock = _this$props.parentLock;
      return templateLock === undefined ? parentLock : templateLock;
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      var innerBlocks = this.props.block.innerBlocks; // only synchronize innerBlocks with template if innerBlocks are empty or a locking all exists

      if (innerBlocks.length === 0 || this.getTemplateLock() === 'all') {
        this.synchronizeBlocksWithTemplate();
      }

      if (this.state.templateInProcess) {
        this.setState({
          templateInProcess: false
        });
      }
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      var _this$props2 = this.props,
          template = _this$props2.template,
          block = _this$props2.block;
      var innerBlocks = block.innerBlocks;
      this.updateNestedSettings(); // only synchronize innerBlocks with template if innerBlocks are empty or a locking all exists

      if (innerBlocks.length === 0 || this.getTemplateLock() === 'all') {
        var hasTemplateChanged = !isEqual(template, prevProps.template);

        if (hasTemplateChanged) {
          this.synchronizeBlocksWithTemplate();
        }
      }
    }
    /**
     * Called on mount or when a mismatch exists between the templates and
     * inner blocks, synchronizes inner blocks with the template, replacing
     * current blocks.
     */

  }, {
    key: "synchronizeBlocksWithTemplate",
    value: function synchronizeBlocksWithTemplate() {
      var _this$props3 = this.props,
          template = _this$props3.template,
          block = _this$props3.block,
          replaceInnerBlocks = _this$props3.replaceInnerBlocks;
      var innerBlocks = block.innerBlocks; // Synchronize with templates. If the next set differs, replace.

      var nextBlocks = _synchronizeBlocksWithTemplate(innerBlocks, template);

      if (!isEqual(nextBlocks, innerBlocks)) {
        replaceInnerBlocks(nextBlocks);
      }
    }
  }, {
    key: "updateNestedSettings",
    value: function updateNestedSettings() {
      var _this$props4 = this.props,
          blockListSettings = _this$props4.blockListSettings,
          allowedBlocks = _this$props4.allowedBlocks,
          updateNestedSettings = _this$props4.updateNestedSettings;
      var newSettings = {
        allowedBlocks: allowedBlocks,
        templateLock: this.getTemplateLock()
      };

      if (!isShallowEqual(blockListSettings, newSettings)) {
        updateNestedSettings(newSettings);
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props5 = this.props,
          clientId = _this$props5.clientId,
          renderAppender = _this$props5.renderAppender,
          renderFooterAppender = _this$props5.renderFooterAppender,
          __experimentalMoverDirection = _this$props5.__experimentalMoverDirection,
          parentWidth = _this$props5.parentWidth,
          horizontal = _this$props5.horizontal,
          contentResizeMode = _this$props5.contentResizeMode,
          contentStyle = _this$props5.contentStyle,
          onAddBlock = _this$props5.onAddBlock,
          onDeleteBlock = _this$props5.onDeleteBlock,
          marginVertical = _this$props5.marginVertical,
          marginHorizontal = _this$props5.marginHorizontal,
          horizontalAlignment = _this$props5.horizontalAlignment;
      var templateInProcess = this.state.templateInProcess;
      return createElement(Fragment, null, !templateInProcess && createElement(BlockList, {
        marginVertical: marginVertical,
        marginHorizontal: marginHorizontal,
        rootClientId: clientId,
        renderAppender: renderAppender,
        renderFooterAppender: renderFooterAppender,
        withFooter: false,
        __experimentalMoverDirection: __experimentalMoverDirection,
        parentWidth: parentWidth,
        horizontalAlignment: horizontalAlignment,
        horizontal: horizontal,
        contentResizeMode: contentResizeMode,
        contentStyle: contentStyle,
        onAddBlock: onAddBlock,
        onDeleteBlock: onDeleteBlock
      }));
    }
  }]);

  return InnerBlocks;
}(Component);

InnerBlocks = compose([withBlockEditContext(function (context) {
  return pick(context, ['clientId']);
}), withSelect(function (select, ownProps) {
  var _select = select('core/block-editor'),
      isBlockSelected = _select.isBlockSelected,
      hasSelectedInnerBlock = _select.hasSelectedInnerBlock,
      getBlock = _select.getBlock,
      getBlockListSettings = _select.getBlockListSettings,
      getBlockRootClientId = _select.getBlockRootClientId,
      getTemplateLock = _select.getTemplateLock;

  var clientId = ownProps.clientId;
  var block = getBlock(clientId);
  var rootClientId = getBlockRootClientId(clientId);
  return {
    block: block,
    blockListSettings: getBlockListSettings(clientId),
    hasOverlay: block.name !== 'core/template' && !isBlockSelected(clientId) && !hasSelectedInnerBlock(clientId, true),
    parentLock: getTemplateLock(rootClientId)
  };
}), withDispatch(function (dispatch, ownProps) {
  var _dispatch = dispatch('core/block-editor'),
      _replaceInnerBlocks = _dispatch.replaceInnerBlocks,
      updateBlockListSettings = _dispatch.updateBlockListSettings;

  var block = ownProps.block,
      clientId = ownProps.clientId,
      _ownProps$templateIns = ownProps.templateInsertUpdatesSelection,
      templateInsertUpdatesSelection = _ownProps$templateIns === void 0 ? true : _ownProps$templateIns;
  return {
    replaceInnerBlocks: function replaceInnerBlocks(blocks) {
      _replaceInnerBlocks(clientId, blocks, block.innerBlocks.length === 0 && templateInsertUpdatesSelection);
    },
    updateNestedSettings: function updateNestedSettings(settings) {
      dispatch(updateBlockListSettings(clientId, settings));
    }
  };
})])(InnerBlocks); // Expose default appender placeholders as components.

InnerBlocks.DefaultBlockAppender = DefaultBlockAppender;
InnerBlocks.ButtonBlockAppender = ButtonBlockAppender;
InnerBlocks.Content = withBlockContentContext(function (_ref) {
  var BlockContent = _ref.BlockContent;
  return createElement(BlockContent, null);
});
/**
 * @see https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/inner-blocks/README.md
 */

export default InnerBlocks;
