// @wordpress/block-editor/build/components/block-navigation/branch.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BlockNavigationListItemFill = exports.BlockNavigationListItemSlot = exports.default = void 0;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _components = require("@wordpress/components");

var _listItem = _interopRequireDefault(require("./list-item"));

var _list = require("./list");

var _block = require("../block-list/block");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var BlockNavigationBranch = function BlockNavigationBranch(_ref) {
  var children = _ref.children,
      props = (0, _objectWithoutProperties2.default)(_ref, ["children"]);

  var _useContext = (0, _element.useContext)(_list.BlockNavigationContext),
      __experimentalWithBlockNavigationSlots = _useContext.__experimentalWithBlockNavigationSlots;

  if (!__experimentalWithBlockNavigationSlots) {
    return (0, _element.createElement)("li", null, (0, _element.createElement)(_listItem.default, props), children);
  }

  return (0, _element.createElement)("li", null, (0, _element.createElement)(BlockNavigationListItemSlot, {
    blockId: props.block.clientId
  }, function (fills) {
    if (!fills.length) {
      return (0, _element.createElement)(_listItem.default, props);
    }

    return _element.Children.map(fills, function (fill) {
      return (0, _element.cloneElement)(fill, _objectSpread({}, props, {}, fill.props));
    });
  }), children);
};

var _default = BlockNavigationBranch;
exports.default = _default;

var listItemSlotName = function listItemSlotName(blockId) {
  return "BlockNavigationList-item-".concat(blockId);
};

var BlockNavigationListItemSlot = function BlockNavigationListItemSlot(_ref2) {
  var blockId = _ref2.blockId,
      props = (0, _objectWithoutProperties2.default)(_ref2, ["blockId"]);
  return (0, _element.createElement)(_components.Slot, (0, _extends2.default)({}, props, {
    name: listItemSlotName(blockId)
  }));
};

exports.BlockNavigationListItemSlot = BlockNavigationListItemSlot;

var BlockNavigationListItemFill = function BlockNavigationListItemFill(props) {
  var _useContext2 = (0, _element.useContext)(_block.BlockListBlockContext),
      clientId = _useContext2.clientId;

  return (0, _element.createElement)(_components.Fill, (0, _extends2.default)({}, props, {
    name: listItemSlotName(clientId)
  }));
};

exports.BlockNavigationListItemFill = BlockNavigationListItemFill;
