// @wordpress/block-editor/build-module/components/block-mover/index.native.js
import { createElement, Fragment } from "@wordpress/element";

/**
 * External dependencies
 */
import { first, last, partial, castArray } from 'lodash';
/**
 * WordPress dependencies
 */

import { ToolbarButton } from '@wordpress/components';
import { __, sprintf } from '@wordpress/i18n';
import { withSelect, withDispatch } from '@wordpress/data';
import { withInstanceId, compose } from '@wordpress/compose';
import { arrowUp, arrowDown, arrowLeft, arrowRight } from '@wordpress/icons';
var horizontalMover = {
  backwardButtonIcon: arrowLeft,
  forwardButtonIcon: arrowRight,
  backwardButtonHint: __('Double tap to move the block to the left'),
  forwardButtonHint: __('Double tap to move the block to the right'),
  firstBlockTitle: __('Move block left'),
  lastBlockTitle: __('Move block right'),

  /* translators: accessibility text. %1: current block position (number). %2: next block position (number) */
  backwardButtonTitle: __('Move block left from position %1$s to position %2$s'),

  /* translators: accessibility text. %1: current block position (number). %2: next block position (number) */
  forwardButtonTitle: __('Move block right from position %1$s to position %2$s')
};
var verticalMover = {
  backwardButtonIcon: arrowUp,
  forwardButtonIcon: arrowDown,
  backwardButtonHint: __('Double tap to move the block up'),
  forwardButtonHint: __('Double tap to move the block down'),
  firstBlockTitle: __('Move block up'),
  lastBlockTitle: __('Move block down'),

  /* translators: accessibility text. %1: current block position (number). %2: next block position (number) */
  backwardButtonTitle: __('Move block up from row %1$s to row %2$s'),

  /* translators: accessibility text. %1: current block position (number). %2: next block position (number) */
  forwardButtonTitle: __('Move block down from row %1$s to row %2$s')
};

var BlockMover = function BlockMover(_ref) {
  var isFirst = _ref.isFirst,
      isLast = _ref.isLast,
      isRTL = _ref.isRTL,
      isLocked = _ref.isLocked,
      onMoveDown = _ref.onMoveDown,
      onMoveUp = _ref.onMoveUp,
      firstIndex = _ref.firstIndex,
      rootClientId = _ref.rootClientId,
      isStackedHorizontally = _ref.isStackedHorizontally;

  var _ref2 = isStackedHorizontally ? horizontalMover : verticalMover,
      backwardButtonIcon = _ref2.backwardButtonIcon,
      forwardButtonIcon = _ref2.forwardButtonIcon,
      backwardButtonHint = _ref2.backwardButtonHint,
      forwardButtonHint = _ref2.forwardButtonHint,
      firstBlockTitle = _ref2.firstBlockTitle,
      lastBlockTitle = _ref2.lastBlockTitle;

  if (isLocked || isFirst && isLast && !rootClientId) {
    return null;
  }

  var switchButtonPropIfRTL = function switchButtonPropIfRTL(isBackwardButton, forwardButtonProp, backwardButtonProp) {
    if (isRTL && isStackedHorizontally) {
      // for RTL and horizontal direction switch prop between forward and backward button
      if (isBackwardButton) {
        return forwardButtonProp; // set forwardButtonProp for backward button
      }

      return backwardButtonProp; // set backwardButtonProp for forward button
    }

    return isBackwardButton ? backwardButtonProp : forwardButtonProp;
  };

  var getMoverButtonTitle = function getMoverButtonTitle(isBackwardButton) {
    var fromIndex = firstIndex + 1; // current position based on index
    // for backwardButton decrease index (move left/up) for forwardButton increase index (move right/down)

    var direction = isBackwardButton ? -1 : 1;
    var toIndex = fromIndex + direction; // position after move

    var _ref3 = isStackedHorizontally ? horizontalMover : verticalMover,
        backwardButtonTitle = _ref3.backwardButtonTitle,
        forwardButtonTitle = _ref3.forwardButtonTitle;

    var buttonTitle = switchButtonPropIfRTL(isBackwardButton, forwardButtonTitle, backwardButtonTitle);
    return sprintf(buttonTitle, fromIndex, toIndex);
  };

  var getArrowIcon = function getArrowIcon(isBackwardButton) {
    return switchButtonPropIfRTL(isBackwardButton, forwardButtonIcon, backwardButtonIcon);
  };

  return createElement(Fragment, null, createElement(ToolbarButton, {
    title: !isFirst ? getMoverButtonTitle(true) : firstBlockTitle,
    isDisabled: isFirst,
    onClick: onMoveUp,
    icon: getArrowIcon(true),
    extraProps: {
      hint: backwardButtonHint
    }
  }), createElement(ToolbarButton, {
    title: !isLast ? getMoverButtonTitle() : lastBlockTitle,
    isDisabled: isLast,
    onClick: onMoveDown,
    icon: getArrowIcon(),
    extraProps: {
      hint: forwardButtonHint
    }
  }));
};

export default compose(withSelect(function (select, _ref4) {
  var clientIds = _ref4.clientIds;

  var _select = select('core/block-editor'),
      getBlockIndex = _select.getBlockIndex,
      getTemplateLock = _select.getTemplateLock,
      getBlockRootClientId = _select.getBlockRootClientId,
      getBlockOrder = _select.getBlockOrder,
      getSettings = _select.getSettings;

  var normalizedClientIds = castArray(clientIds);
  var firstClientId = first(normalizedClientIds);
  var rootClientId = getBlockRootClientId(firstClientId);
  var blockOrder = getBlockOrder(rootClientId);
  var firstIndex = getBlockIndex(firstClientId, rootClientId);
  var lastIndex = getBlockIndex(last(normalizedClientIds), rootClientId);
  return {
    firstIndex: firstIndex,
    isFirst: firstIndex === 0,
    isLast: lastIndex === blockOrder.length - 1,
    isRTL: getSettings().isRTL,
    isLocked: getTemplateLock(rootClientId) === 'all',
    rootClientId: rootClientId
  };
}), withDispatch(function (dispatch, _ref5) {
  var clientIds = _ref5.clientIds,
      rootClientId = _ref5.rootClientId;

  var _dispatch = dispatch('core/block-editor'),
      moveBlocksDown = _dispatch.moveBlocksDown,
      moveBlocksUp = _dispatch.moveBlocksUp;

  return {
    onMoveDown: partial(moveBlocksDown, clientIds, rootClientId),
    onMoveUp: partial(moveBlocksUp, clientIds, rootClientId)
  };
}), withInstanceId)(BlockMover);
